src/main/java/org/opensearch/ad/indices/AnomalyDetectionIndices.java

```diff
@@ -603,34 +630,99 @@ private void deleteIndexIteration(String[] toDelete) {
         }
     }
 
+    public void update() {
+        if ((allMappingUpdated && allSettingUpdated) || updateRunningTimes >= maxUpdateRunningTimes || updateRunning.get()) {
+            return;
+        }
+        updateRunning.set(true);
+        updateRunningTimes++;
+
+        // set updateRunning to false when both updateMappingIfNecessary and updateSettingIfNecessary
+        // stop running
+        final GroupedActionListener<Void> groupListeneer = new GroupedActionListener<>(
+            ActionListener.wrap(r -> updateRunning.set(false), exception -> {
+                updateRunning.set(false);
+                logger.error("Fail to update AD indices", exception);
+            }),
+            // 2 since we need both updateMappingIfNecessary and updateSettingIfNecessary to return
+            // before setting updateRunning to false
+            2
+        );
+
+        updateMappingIfNecessary(groupListeneer);
+        updateSettingIfNecessary(groupListeneer);
+    }
+
+    private void updateSettingIfNecessary(GroupedActionListener<Void> delegateListeneer) {
+        if (allSettingUpdated) {
+            delegateListeneer.onResponse(null);
+            return;
+        }
+
+        List<ADIndex> updates = new ArrayList<>();
+        for (ADIndex index : ADIndex.values()) {
+            Boolean updated = indexStates.computeIfAbsent(index, IndexState::new).settingUpToDate;
+            if (Boolean.FALSE.equals(updated)) {
+                updates.add(index);
+            }
+        }
+        if (updates.size() == 0) {
+            allSettingUpdated = true;
+            delegateListeneer.onResponse(null);
+            return;
+        }
+
+        final GroupedActionListener<Void> conglomerateListeneer = new GroupedActionListener<>(
+            ActionListener.wrap(r -> delegateListeneer.onResponse(null), exception -> {
+                delegateListeneer.onResponse(null);
+                logger.error("Fail to update AD indices' mappings", exception);
+            }),
+            updates.size()
+        );
+        for (ADIndex adIndex : updates) {
+            logger.info(new ParameterizedMessage("Check [{}]'s setting", adIndex.getIndexName()));
+            switch (adIndex) {
+                case JOB:
+                    updateJobIndexSettingIfNecessary(indexStates.computeIfAbsent(adIndex, IndexState::new), conglomerateListeneer);
+                    break;
+                default:
+                    // we don't have settings to update for other indices
+                    IndexState indexState = indexStates.computeIfAbsent(adIndex, IndexState::new);
+                    indexState.settingUpToDate = true;
+                    logger.info(new ParameterizedMessage("Mark [{}]'s setting up-to-date", adIndex.getIndexName()));
+                    conglomerateListeneer.onResponse(null);
```

Comment:
`delegateListeneer` in this method is `GroupedActionListener` with group size 2. 
`conglomerateListeneer` is `GroupedActionListener` with group size `updates.size()`. 
Do these two group listeners have same group size?

Reply:
Oh, never mind, misunderstood the group size, should be ok, as the `conglomerateListeneer` will only call `delegateListeneer` once all updates returns.